components/filter/filter.js

```diff
@@ -0,0 +1,259 @@
+				extend-separators>
+
+				${items.map(item => html`
+					<d2l-list-item
+						key="${item.key}"
+						selectable
+						?selected="${item.selected}"
+						slim>
+
+						<div class="d2l-filter-dimension-value-text">${item.text}</div>
+					</d2l-list-item>
+				`)}
+			</d2l-list>
+		`;
+	}
+
+	_dispatchChangeEvent(eventDetail) {
+		this.dispatchEvent(new CustomEvent('d2l-filter-change', { bubbles: false, composed: false, detail: eventDetail }));
+	}
+
+	_handleChangeListDimension(e) {
+		const dimensionKey = e.composedPath()[0].parentNode.id;
+		const valueKey = e.detail.key;
+		const selected = e.detail.selected;
+
+		// Update the corresponding d2l-filter-dimension-value to keep them in sync
+		const dimension = this._dimensions.get(dimensionKey);
+		const slot = dimension.shadowRoot.querySelector('slot');
+		const items = slot.assignedNodes().filter(node => node.nodeType ===  Node.ELEMENT_NODE);
+		const item = items.find(item => item.key === valueKey);
+		item.selected = selected;
+
+		this._dispatchChangeEvent({ dimension: dimensionKey, value: { key: valueKey, selected: selected } });
+	}
+
+	_handleDataChangeListDimension(e) {
+		const dimension = this.shadowRoot.getElementById(e.target.key);
+		const changes = e.detail.changes;
+		if (changes.has('text')) {
+			const menuItem = dimension.parentNode;
+			if (menuItem.text !== changes.get('text')) {
+				this.requestUpdate();
+				return;
+			}
+		}
+	}
+
+	_handleDataChangeListDimensionValue(e) {
+		const dimension = this.shadowRoot.getElementById(e.target.parentNode.key);
+		const value = dimension.querySelector(`[key="${e.target.key}"]`);
+		const changes = e.detail.changes;
+
+		if (changes.has('text')) {
+			const text = value.querySelector('.d2l-filter-dimension-value-text');
+			if (text.innerText !== changes.get('text')) {
+				text.innerText = changes.get('text');
+			}
+		}
+		if (changes.has('selected')) {
+			if (value.selected !== changes.get('selected')) {
+				value.selected = changes.get('selected');
+			}
+		}
+	}
+
+	_handleDimensionHide() {
+		this.shadowRoot.querySelector(`#${this._activeDimensionKey}`).hide();
+		this._activeDimensionKey = null;
+	}
+
+	_handleDimensionShowComplete() {
+		const dimension = this.shadowRoot.querySelector(`#${this._activeDimensionKey}`);
+		const focusable = getFirstFocusableDescendant(dimension);
+		if (focusable) {
+			focusable.focus();
+		}
+	}
+
+	_handleDimensionShowStart(e) {
+		this._activeDimensionKey = e.detail.sourceView.id;
+	}
+
+	_handleDropdownClose(e) {
+		this._activeDimensionKey = null;
+		this._stopPropagation(e);
+	}
+
+	_handleSlotChange() {
+		const nodes = this._dimensionsSlot.assignedNodes();
+		const filteredNodes = nodes.filter((node) => {
+			const isNode = node.nodeType === Node.ELEMENT_NODE && node.tagName.toLowerCase() !== 'template';
+			return isNode;
+		});
+
+		this._dimensions = new Map(filteredNodes.map(node => [node.key, node]));
```

Comment:
I convert this to a map so that when I do a number of lookups above it's cleaner and faster, but not sure if it's worth it overall